File: Formula/f2format.rb
class F2format < Formula
  include Language::Python::Virtualenv

  desc "Backport compiler for Python 3.6 f-string literals"
  homepage "https://github.com/pybpc/f2format#f2format"
  url "https://github.com/pybpc/f2format/archive/v0.8.7.tar.gz"
  sha256 "d5558cd419c8d46bdc958064cb97f963d1ea793866414c025906ec15033512ed"

  head "https://github.com/pybpc/f2format.git", branch: "master"

  depends_on "homebrew/core/python@3.9"

  resource "bpc-utils" do
    url "https://files.pythonhosted.org/packages/c8/b7/8131c9884d2d79c5b68d6ad8856aa7e7d89543a4dda7ff5fa0abc49e6f80/bpc-utils-0.10.1.tar.gz"
    sha256 "0954282e947e712152b6396f7b04487c28e2cdfef7e4538453cb13662faa7d33"
  end

  resource "parso" do
    url "https://files.pythonhosted.org/packages/a2/0e/41f0cca4b85a6ea74d66d2226a7cda8e41206a624f5b330b958ef48e2e52/parso-0.8.3.tar.gz"
    sha256 "8c07be290bb59f03588915921e29e8a50002acaf2cdc5fa0e0114f91709fafa0"
  end

  resource "typing-extensions" do
    url "https://files.pythonhosted.org/packages/16/3a/0d26ce356c7465a19c9ea8814b960f8a36c3b0d07c323176620b7b483e44/typing_extensions-4.10.0.tar.gz"
    sha256 "b0abd7c89e8fb96f98db18d86106ff1d90ab692004eb746cf6eda2682f91b3cb"
  end

  resource "parso" do
    url "https://files.pythonhosted.org/packages/a2/0e/41f0cca4b85a6ea74d66d2226a7cda8e41206a624f5b330b958ef48e2e52/parso-0.8.3.tar.gz"
    sha256 "8c07be290bb59f03588915921e29e8a50002acaf2cdc5fa0e0114f91709fafa0"
  end

  resource "tbtrim" do
    url "https://files.pythonhosted.org/packages/85/62/89756f5d2d61691591c4293fd4cc1fbb3aab1466251c7319fe60dd08fb27/tbtrim-0.3.1.tar.gz"
    sha256 "b6285ac02e9a7b78fab97de65668fe2def0f5d8783b0e0dfcb4b7c0a635b3d11"
  end

  def install
    virtualenv_install_with_resources
    man1.install "share/f2format.1"
    bash_completion.install "share/f2format.bash-completion"
  end

  test do
    (testpath/"test.py").write <<~EOS
      var = f'foo{(1+2)*3:>5}bar{"a", "b"!r}boo'
    EOS

    std_output = <<~EOS
      var = 'foo{:>5}bar{!r}boo'.format((1+2)*3, ("a", "b"))
    EOS

    system bin/"f2format", "--no-archive", "test.py"
    assert_match std_output, shell_output("cat test.py")
  end
end

File: Generator/scripts/f2format.py
# -*- coding: utf-8 -*-

import hashlib
import os
import re
import subprocess  # nosec: B404
from typing import TYPE_CHECKING

import requests

# with open(os.path.expanduser('~/GitHub/f2format/f2format.py'), 'r') as file:
#     for line in file:
#         match = re.match(r"^__version__ = '(.*)'", line)
#         if match is None:
#             continue
#         VERSION = match.groups()[0]
#         break
# # print(VERSION)
if TYPE_CHECKING:
    VERSION: str

subprocess.check_call(['pip', 'install', 'bpc-f2format'])  # nosec: B603 B607

for line in subprocess.check_output(['pip', 'freeze']).decode().splitlines():  # nosec
    match = re.match(r"bpc-f2format==(.*)", line, re.IGNORECASE)
    if match is not None:
        VERSION = match.groups()[0]

F2FORMAT_URL = f'https://github.com/pybpc/f2format/archive/v{VERSION}.tar.gz'
F2FORMAT_SHA = hashlib.sha256(requests.get(F2FORMAT_URL).content).hexdigest()
# print(F2FORMAT_URL)
# print(F2FORMAT_SHA)

PARSO = subprocess.check_output(['poet', 'parso']).decode().strip()  # nosec: B603 B607
TBTRIM = subprocess.check_output(['poet', 'tbtrim']).decode().strip()  # nosec: B603 B607
BPC_UTILS = subprocess.check_output(['poet', 'bpc-utils']).decode().strip()  # nosec: B603 B607
# print(PARSO)
# print(TBTRIM)

match = re.match(r'([0-9.]+)\.post([0-9]+)', VERSION)
if match is None:
    F2FORMAT = (f'url "{F2FORMAT_URL}"\n'
                f'  sha256 "{F2FORMAT_SHA}"')
else:
    version, subversion = match.groups()
    revision = chr(96 + int(subversion))  # ord('a') -> 97
    F2FORMAT = (f'url "{F2FORMAT_URL}"\n'
                f'  version "{version}{revision}"\n'
                f'  sha256 "{F2FORMAT_SHA}"')


FORMULA = f'''\
class F2format < Formula
  include Language::Python::Virtualenv

  desc "Backport compiler for Python 3.6 f-string literals"
  homepage "https://github.com/pybpc/f2format#f2format"
  {F2FORMAT}

  head "https://github.com/pybpc/f2format.git", branch: "master"

  depends_on "homebrew/core/python@3.9"

  {BPC_UTILS}

  {PARSO}

  {TBTRIM}

  def install
    virtualenv_install_with_resources
    man1.install "share/f2format.1"
    bash_completion.install "share/f2format.bash-completion"
  end

  test do
    (testpath/"test.py").write <<~EOS
      var = f'foo{{(1+2)*3:>5}}bar{{"a", "b"!r}}boo'
    EOS

    std_output = <<~EOS
      var = 'foo{{:>5}}bar{{!r}}boo'.format((1+2)*3, ("a", "b"))
    EOS

    system bin/"f2format", "--no-archive", "test.py"
    assert_match std_output, shell_output("cat test.py")
  end
end
'''

if os.path.basename(__file__) == 'setup-formula.py':
    repo_root = subprocess.check_output(['brew', '--repository', 'jarryshaw/tap'],  # nosec: B603 B607
                                        encoding='utf-8').strip()
    formula = os.path.join(repo_root, 'Formula', 'f2format.rb')
else:
    formula = os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..', 'Formula',
                           f'{os.path.splitext(os.path.basename(__file__))[0]}.rb')
with open(formula, 'w', encoding='utf-8') as file:
    file.write(FORMULA)

subprocess.check_call(['pip-autoremove', '-y', 'bpc-f2format'])  # nosec: B603 B607
